File: 4_Behavioral_Design_Patterns/4.09_Strategy/Python_Snippet/Strategy.py
from abc import ABC,abstractmethod

# Strategy Interface
class RoutingStrategy(ABC):
    @abstractmethod
    def calculate_route(self, origin, destination):
        pass

# Concrete Strategy: ShortestDistanceRouting
class ShortestDistanceRouting(RoutingStrategy):
    def calculate_route(self, origin, destination):
        # Logic to calculate the shortest distance route
        print(f"Calculating shortest distance route from {origin} to {destination}")

# Concrete Strategy: FastestRouteRouting
class FastestRouteRouting(RoutingStrategy):
    def calculate_route(self, origin, destination):
        # Logic to calculate the fastest route
        print(f"Calculating fastest route from {origin} to {destination}")

# Concrete Strategy: AvoidTollRoadsRouting
class AvoidTollRoadsRouting(RoutingStrategy):
    def calculate_route(self, origin, destination):
        # Logic to calculate a route avoiding toll roads
        print(f"Calculating route from {origin} to {destination} while avoiding toll roads")

# Context: NavigationApp
class NavigationApp:
    def __init__(self, routing_strategy):
        self.routing_strategy = routing_strategy

    def set_routing_strategy(self, routing_strategy):
        self.routing_strategy = routing_strategy

    def calculate_route(self, origin, destination):
        self.routing_strategy.calculate_route(origin, destination)

# Usage
if __name__ == "__main__":
    origin = "A"
    destination = "B"

    navigation_app = NavigationApp(ShortestDistanceRouting())
    navigation_app.calculate_route(origin, destination)
    # Output: Calculating shortest distance route from A to B

    navigation_app.set_routing_strategy(FastestRouteRouting())
    navigation_app.calculate_route(origin, destination)
    # Output: Calculating fastest route from A to B

    navigation_app.set_routing_strategy(AvoidTollRoadsRouting())
    navigation_app.calculate_route(origin, destination)
    # Output: Calculating route from A to B while avoiding toll roads

File: 4_Behavioral_Design_Patterns/4.09_Strategy/Java_Snippet/Strategy.java
// Strategy Interface
interface RoutingStrategy {
    void calculateRoute(String origin, String destination);
}

// Concrete Strategy: ShortestDistanceRouting
class ShortestDistanceRouting implements RoutingStrategy {
    @Override
    public void calculateRoute(String origin, String destination) {
        // Logic to calculate the shortest distance route
        System.out.println("Calculating shortest distance route from " + origin + " to " + destination);
    }
}

// Concrete Strategy: FastestRouteRouting
class FastestRouteRouting implements RoutingStrategy {
    @Override
    public void calculateRoute(String origin, String destination) {
        // Logic to calculate the fastest route
        System.out.println("Calculating fastest route from " + origin + " to " + destination);
    }
}

// Concrete Strategy: AvoidTollRoadsRouting
class AvoidTollRoadsRouting implements RoutingStrategy {
    @Override
    public void calculateRoute(String origin, String destination) {
        // Logic to calculate a route avoiding toll roads
        System.out.println("Calculating route from " + origin + " to " + destination + " while avoiding toll roads");
    }
}

// Context: NavigationApp
class NavigationApp {
    private RoutingStrategy routingStrategy;

    public NavigationApp(RoutingStrategy routingStrategy) {
        this.routingStrategy = routingStrategy;
    }

    public void setRoutingStrategy(RoutingStrategy routingStrategy) {
        this.routingStrategy = routingStrategy;
    }

    public void calculateRoute(String origin, String destination) {
        this.routingStrategy.calculateRoute(origin, destination);
    }
}

// Usage
public class Strategy {
    public static void main(String[] args) {
        String origin = "A";
        String destination = "B";

        NavigationApp navigationApp = new NavigationApp(new ShortestDistanceRouting());
        navigationApp.calculateRoute(origin, destination);
        // Output: Calculating shortest distance route from A to B

        navigationApp.setRoutingStrategy(new FastestRouteRouting());
        navigationApp.calculateRoute(origin, destination);
        // Output: Calculating fastest route from A to B

        navigationApp.setRoutingStrategy(new AvoidTollRoadsRouting());
        navigationApp.calculateRoute(origin, destination);
        // Output: Calculating route from A to B while avoiding toll roads
    }
}
